diff --git a/include/hardware/audio.h b/include/hardware/audio.h
index 22e7419..a73c10d 100644
--- a/include/hardware/audio.h
+++ b/include/hardware/audio.h
@@ -530,6 +530,8 @@ struct audio_hw_device {
     /** set the audio volume of a voice call. Range is between 0.0 and 1.0 */
     int (*set_voice_volume)(struct audio_hw_device *dev, float volume);
 
+    int (*set_fm_volume)(float volume);
+
     /**
      * set the audio volume for all audio activities other than voice call.
      * Range between 0.0 and 1.0. If any value other than 0 is returned,
