diff --git a/rootdir/init.rc b/rootdir/init.rc
index cfcab62..2f9c428 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -711,15 +711,6 @@ service installd /system/bin/installd
     class main
     socket installd stream 600 system system
 
-service flash_recovery /system/bin/install-recovery.sh
-    class main
-    oneshot
-    disabled
-
-# update recovery if enabled
-on property:persist.sys.recovery_update=true
-    start flash_recovery
-
 service racoon /system/bin/racoon
     class main
     socket racoon stream 600 system system
